Split RW and RO Layer locks per method

saschagrunert opened the pull request:
The new storage options `ReadOnlyLayers` indicates that the
layer store should be locked for read only access. This allows use cases
like listing container images while pushing an image to a remote
location in parallel.

Closes #420

saschagrunert commented:
This is now ready for a first review, I tried to implement with as low overhead as possible.

rhatdan commented:
Tests are failing?

saschagrunert commented:
> Tests are failing?

Yes, thanks for the hint. 

My main concern right now is that we give the responsibility over "locking the right way" to the consumer of the API. But since the option is off per default and not serialized into the config, it should be fine...

rhatdan commented:
LGTM
Need @nalind @vrothberg to review.  Since they have played with locks a lot more then I have.
@mheon PTAL

vrothberg commented:
Aren't we having a meeting for this topic on Friday?

saschagrunert commented:
> Aren't we having a meeting for this topic on Friday?

The intention of the meeting was more about locking enhancements per layer, see: https://github.com/containers/storage/issues/421

vrothberg commented on layers.go:
```diff
@@ -486,7 +486,11 @@ func newLayerStore(rundir string, layerdir string, driver drivers.Driver, uidMap
+		lockfile.RLock()
+	} else {
+		lockfile.Lock()
+	}
```
I wonder why we to lock here at all. We're only calling public APIs; `GetLockfile` serializes own its own and `(*layerStore) Load` is locking as well. I think we can ditch the locking here entirely; @nalind PTAL.

vrothberg replied:
There's also `newROLayerStore`: Why not use that?

vrothberg commented on store.go:
```diff
@@ -3139,7 +3148,11 @@ func (s *store) Shutdown(force bool) ([]string, error) {
 
-	rlstore.Lock()
+	if s.readOnlyLayers {
+		rlstore.RLock()
```
A reader lock doesn't look right here as we're unmounting layers.

saschagrunert commented on store.go:
```diff
@@ -3119,10 +3180,26 @@ func (s *store) FromContainerRunDirectory(id, file string) ([]byte, error) {
+	err := s.graphDriver.Cleanup()
+	s.graphLock.Touch()
+	return err
+}
```
TODO: Check if this is the right cleanup strategy.

saschagrunert commented:
Ready for a new round of review, @vrothberg @nalind @rhatdan PTAL

vrothberg commented:
> Ready for a new round of review, @vrothberg @nalind @rhatdan PTAL

Can you explain what the code is intended to be doing now (and ideally update the commit message)? Why can't we drop the locks (see https://github.com/containers/storage/pull/473#discussion_r352513599) causing parts of the problem?

saschagrunert commented:
> Why can't we drop the locks (see [#473 (comment)](https://github.com/containers/storage/pull/473#discussion_r352513599)) causing parts of the problem?

We can, but we change the behavior of the cleanup on init. I don't think that this is harmful and created a dedicated PR for that, since it is not directly related to this topic: https://github.com/containers/storage/pull/493

rhatdan commented:
@saschagrunert @vrothberg @nalind Bringing this back up for discussion.  Where are we on this now?  Should we go forward with this?

rhatdan commented:
@saschagrunert @vrothberg PING...

vrothberg commented:
We should reevaluate how far #493 brought us.

saschagrunert commented:
Yes some details have change but the main approach would be the same. I ping you here when this is ready for review.

saschagrunert commented:
https://github.com/containers/libpod/pull/4660 is now working, PTAL @rhatdan @vrothberg @nalind

rhatdan commented:
@vrothberg @nalind PTAL

rhatdan commented:
My understanding is, right now if to users of the storage want to use it at the same time, the second one in, needs to wait til the first is done.  In the new design, if both users want to READ, then they are allowed in, and only writers would block.  So a writer can not write while a reader exists.  And readers can not read when a writer holds the lock.

I guess there is a chance that writers would get frozen out if a continuous stream of readers come in.

vrothberg commented:
> My understanding is, right now if to users of the storage want to use it at the same time, the second one in, needs to wait til the first is done. In the new design, if both users want to READ, then they are allowed in, and only writers would block. So a writer can not write while a reader exists. And readers can not read when a writer holds the lock.

The main issue IMHO is that we decide the READ or WRITE access when creating the store and try to enforce that. So we are not fixing/optimizing the code paths but change the nature of the locks altogether. This only works for short-lived store objects (see https://github.com/containers/libpod/pull/4660/files) and is not generally useful for CRI-O (or the `podman system service`), for instance, where the store is not created for each request but is a long-lived object.

Assume I want to access the store as a reader: I have to create a new object, then do my work and subsequent writing accesses would require to create a new object. I would prefer to make c/storage smart enough to figure out when we need to acquire the locks as a reader or writer (and further localize the bottlenecks). It really feels like a symptomatic patch while the underlying bottlenecks in c/storage remain.

rhatdan commented:
Ok so Nalin and I talked about this further, and the problem is each Runtime when it starts takes a Read/Write lock, even if they do not need it.  So when you are pushing the image, you only have a read/lock, but when `podman images` (buildah images) starts it wants to create a read/write lock initially and then waits until `podman push` is done.  Wouldn't a better solution be to take a read/only lock and then when I call into a storage method that requires read/write, the method checks if the lock is read/write and it takes it, if not it ReOpens the lock or Recreates the lock as a read/write lock and then locks it.

vrothberg commented:
> Wouldn't a better solution be to take a read/only lock and then when I call into a storage method that requires read/write, the method checks if the lock is read/write and it takes it, if not it ReOpens the lock or Recreates the lock as a read/write lock and then locks it.

I preferred this ^ approach, yes. It's transparent to consumers of c/storage and seems to tackle the root cause.

saschagrunert commented:
> Wouldn't a better solution be to take a read/only lock and then when I call into a storage method that requires read/write, the method checks if the lock is read/write and it takes it, if not it ReOpens the lock or Recreates the lock as a read/write lock and then locks it.

Alright, I'm moving forward with this approach right now. I split-up the locks (added a new RO lockfile) to not clash with the RW lock and moved the read/write logic into the dedicated methods. This works for now, let's see how podman works with it.

saschagrunert commented:
Ready for review, again.

rhatdan commented:
Need @nalind and @vrothberg to approve.

rhatdan commented:
@saschagrunert Only issue, which I don't think we can solve or should solve, is 
```
$(while 1; do podman images; done)
podman pull foobar
```
Podman pull would never succeed, correct?

vrothberg commented on pkg/lockfile/lockfile_unix.go:
```diff
@@ -35,7 +35,7 @@ type lockfile struct {
 func openLock(path string, ro bool) (int, error) {
 	if ro {
-		return unix.Open(path, os.O_RDONLY, 0)
+		return unix.Open(path, os.O_RDONLY|os.O_CREATE, 0)
```
@nalind, do you have any concerns with that change?

saschagrunert replied:
[The main reason behind that change is to be able to initially create the lock if read-only.]

nalind replied:
Yeah, this is fine (didn't know that O_RDONLY | O_CREAT was a thing, so that's neat).

saschagrunert replied:
Awesome, thanks for the clarification.

vrothberg commented on store.go:
```diff
@@ -325,6 +325,9 @@ type Store interface {
 	Layers() ([]Layer, error)
 
+	// LayersCleanedUp returns the layers while cleaning up duplicates before.
+	LayersCleanedUp() ([]Layer, error)
```
Nit: `DeduplicateLayer()` might be more expressive

saschagrunert replied:
Yes, I change that name :+1:

vrothberg commented on store.go:
```diff
@@ -738,39 +745,81 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
 }
 
-// LayerStore obtains and returns a handle to the writeable layer store object
+// RWLayerStore obtains and returns a handle to the writeable layer store object
```
That's an API break which would bump us to v2.0.0 (forcing us into the version prefixing of go modules). I think we should keep `LayerStore` around and make it call `RWLayerStore`.

saschagrunert replied:
Yes, that makes sense. I change it as suggested.

saschagrunert commented:
> @saschagrunert Only issue, which I don't think we can solve or should solve, is
> 
> ```
> $(while 1; do podman images; done)
> podman pull foobar
> ```
> 
> Podman pull would never succeed, correct?

Hm, on my machine that works in parallel, but the only think which blocks is the `Storing signatures` which write-locks the whole storage. This can be solved by per-layer locking. :+1:

vrothberg commented on store.go:
```diff
@@ -738,39 +745,81 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
+	return s.roLayerStore, nil
+}
+
+func (s *store) getLayerPaths() (rlpath, glpath string, driver drivers.Driver, err error) {
```
Please add a comment.

vrothberg commented on store.go:
```diff
@@ -738,39 +745,81 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
+	return rlpath, glpath, driver, nil
+}
+
+func (s *store) invalidateOnTimeout() {
```
Nit: it's not really a timeout (as there's no timer) but a refresh.

vrothberg replied:
Please add a comment.

vrothberg commented on store.go:
```diff
@@ -2605,7 +2653,7 @@ func (s *store) DiffSize(from, to string) (int64, error) {
 
 func (s *store) Diff(from, to string, options *DiffOptions) (io.ReadCloser, error) {
-	lstore, err := s.LayerStore()
+	lstore, err := s.RWLayerStore()
```
I think that could be RO as we're not changing state.

saschagrunert replied:
Let me check that via podman.

saschagrunert replied:
Nah it does not seem to work: https://cirrus-ci.com/task/5661436064038912

Let me revert that change.

nalind replied:
Yeah, `Diff()` currently requires a write lock because we need to mount filesystems for the sake of drivers that use naive diff, and I expect `Changes()` does, too.  If we pushed the locking down into individual drivers, we could remove that requirement for drivers like overlay that use a native diff.  We'll probably have to do that, since we want to be able to push the contents of those layers to registries when those layers were part of a base image in a build.

vrothberg commented on store.go:
```diff
@@ -2750,7 +2798,7 @@ func (s *store) LayerParentOwners(id string) ([]int, []int, error) {
 
 func (s *store) ContainerParentOwners(id string) ([]int, []int, error) {
-	rlstore, err := s.LayerStore()
+	rlstore, err := s.RWLayerStore()
```
This could be RO as we're not changing state.

saschagrunert replied:
`ParentOwners` is currently not part of the `ROLayerStore` interface because we would require the mount information for that method.

nalind replied:
FWIW, the `...ParentOwners()` family of methods was an experiment that we ended up not using, so if/when we want to break API, we should junk them.

saschagrunert replied:
I can follow-up on that in another PR :)

vrothberg commented:
> > @saschagrunert Only issue, which I don't think we can solve or should solve, is
> > ```
> > $(while 1; do podman images; done)
> > podman pull foobar
> > ```
> > 
> > 
> > Podman pull would never succeed, correct?
> 
> Hm, on my machine that works in parallel, but the only think which blocks is the `Storing signatures` which write-locks the whole storage. This can be solved by per-layer locking. +1

Right. Pulling is a writing operation. But speeding this up is something we should do in containers/image and make it transparent to c/storage. https://github.com/containers/image/pull/611 is doing that (among too many other things).

Parallel pushes should be performing much better now.

rhatdan commented on store.go:
```diff
@@ -2782,8 +2841,8 @@ func (s *store) ContainerParentOwners(id string) ([]int, []int, error) {
 
-func (s *store) Layers() ([]Layer, error) {
-	lstore, err := s.LayerStore()
+func (s *store) DeduplicateLayers() ([]Layer, error) {
```
Why isn't this called AllLayers() or RWLayers?
This name is very confusing.

saschagrunert replied:
I renamed it to `RWLayers` :+1:

rhatdan commented on cmd/containers-storage/layers.go:
```diff
@@ -9,10 +9,21 @@ import (
+		layers []storage.Layer
+		err    error
+	)
+	if cleanup {
```
Why is this called cleanup?  It looks like you are only getting RWLayers and printing them?  Not sure what cleanup means in this case?

saschagrunert replied:
I think we can remove the `cleanup` flag and stick to the `RWLayers`-only approach.

rhatdan commented on store.go:
```diff
@@ -325,6 +325,9 @@ type Store interface {
 	// Layers returns a list of the currently known layers.
 	Layers() ([]Layer, error)
 
+	// DeduplicateLayers returns the layers while cleaning up duplicates before.
```
Comment makes no sense.

saschagrunert replied:
Fixed it based on the new function name

nalind commented on layers.go:
```diff
@@ -523,7 +523,7 @@ func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Dri
 
 func newROLayerStore(rundir string, layerdir string, driver drivers.Driver) (ROLayerStore, error) {
-	lockfile, err := GetROLockfile(filepath.Join(layerdir, "layers.lock"))
+	lockfile, err := GetROLockfile(filepath.Join(layerdir, "layers-ro.lock"))
```
Since it's now using a different lock file, how does a read-only store detect that the contents that are protected by the lock have been modified by another process that presumably took a write lock?

saschagrunert replied:
I see this might be a problem. What I now did in https://github.com/containers/storage/pull/473/commits/9941e40a94d55291e24edfd9ea472f1b4d6718ee is adding an additional check before actually creating the store. Do you think we have to add the same pre-check to the `newLayerStore` method as well?

nalind replied:
`newLayerStore` and `newROLayerStore` alway calls `Load()`, so they're fine.  A long-running process isn't necessarily ever going to refresh the read-only store's contents after that, which I think is going to be a problem.

saschagrunert replied:
So do you think we can solve this by acquiring a `layers.lock` and `layers-ro.lock` together on read-write?

nalind replied:
Yes, acquiring both locks, in an order consistent with other cases where both can be acquired at the same time, should avoid this.  I think doing so brings back the lock contention that you're trying to avoid by using a different lock, though, and I'm not sure what to do about that.

saschagrunert replied:
I think if we adapt the locker implementation then we could solve the problem. I now added a more generic lockfile creation function `GetLockfileRWRO()`, which is able to change the locker type (read-only or not) during its usage.

Let's see what the tests say.

rhatdan commented:
LGTM
Needs @nalind and @saschagrunert to agree on solution.

vrothberg commented on pkg/lockfile/lockfile_unix.go:
```diff
@@ -253,3 +253,13 @@ func (l *lockfile) TouchedSince(when time.Time) bool {
 }
+
+// SetReadOnly can be used to set the locker either to `read-only` or `read-write`
+func (l *lockfile) SetReadOnly(ro bool) {
```
It seems odd that `SetReadOnly` can change lock to a writer. Please rename to `LockType(...)` or something similarly expressive.

saschagrunert commented on layers.go:
```diff
@@ -1393,18 +1399,30 @@ func (r *layerStore) LayersByUncompressedDigest(d digest.Digest) ([]Layer, error
+	if r.rolockfile != nil {
+		r.rolockfile.Unlock()
+	}
 }
```
@vrothberg @nalind tried another approach with a separate `r.rolockfile` which is additionally set when layer store is locked read and write. I'm not sure if we can solve possible locking issues with that :thinking:

rhatdan commented:
@kolyshkin PTAL

vrothberg commented:
> Is there any caller trying acquire `layers-ro.lock` as a reader-writer? If so, then the name should be changed as it's not read-only. If not, then _all callers_ would acquire and access it as a reader rendering it a NOP.
> 
> @saschagrunert, can you elaborate a bit what `layers-ro.lock` is intended to do?

Ping.

saschagrunert commented:
> > Is there any caller trying acquire `layers-ro.lock` as a reader-writer? If so, then the name should be changed as it's not read-only. If not, then _all callers_ would acquire and access it as a reader rendering it a NOP.
> > @saschagrunert, can you elaborate a bit what `layers-ro.lock` is intended to do?
> 
> Ping.

Ah, sorry for the late reply. So to sum it up, the main intention is to solve the issue mentioned by @nalind:

> Since it's now using a different lock file, how does a read-only store detect that the contents that are protected by the lock have been modified by another process that presumably took a write lock?
>  A long-running process isn't necessarily ever going to refresh the read-only store's contents after that, which I think is going to be a problem.

The lock implementation is currently not flexible enough to use an initiated read-only lockfile as read-write later on (and the other way around) so my initial approach was to add a third lockfile type which supports that scenario.

The other implementation approach (the current one) was solving the issue via 2 lockfiles, whereas the write-able store also locks the read-only one when acquiring its write-lock.

saschagrunert commented:
Ping :point_up:

vrothberg commented:
> The lock implementation is currently not flexible enough to use an initiated read-only lockfile as read-write later on (and the other way around) so my initial approach was to add a third lockfile type which supports that scenario.

If you need a lock which can be both, a reader and a writer, then you need to pick a writer lock (non-read-only) which supports write and read acquisitions. Read only really means read only in this implementation but ,in general, a lock can change type at anytime.

> The other implementation approach (the current one) was solving the issue via 2 lockfiles, whereas the write-able store also locks the read-only one when acquiring its write-lock.
> 
> I think the current implementation makes no sense from a locking perspective… Do you think we could solve the issue by creating 2 RW locks, which:
> 
>     * are both locked write-able when using the read-write store

Yes. Both need to be acquired as writers to prevent readers from getting old data or even from being corrupted.

>     * are used as write-able and read-only when using the read-only store

If one is acquired as a writer, wouldn't that code path be exclusive and not allow for parallel reads?

Can't we use the existing locks and acquire them as writer/reader when needed?

saschagrunert commented:
> Can't we use the existing locks and acquire them as writer/reader when needed?

Yes, I think my latest implementation approach would work like this. PTAL.

PS: I think travis lost the connection to report the results here and we have to re-connect somehow.

vrothberg commented on layers.go:
```diff
@@ -491,7 +491,7 @@ func (r *layerStore) saveMounts() error {
 }
 
-func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Driver) (LayerStore, error) {
+func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Driver, withMountsLock bool) (LayerStore, error) {
```
Please add a comment explaining the implications of `withMountsLock`. A comment at the definition of `layerStore` would be great as well, so that IDE's and godef will help understanding it.

saschagrunert replied:
Added comment as suggested :+1:

vrothberg commented on layers.go:
```diff
@@ -1431,7 +1413,7 @@ func (r *layerStore) Modified() (bool, error) {
 
 func (r *layerStore) IsReadWrite() bool {
-	return r.lockfile.IsReadWrite()
+	return r.lockfile.IsReadWrite() && r.mountsLockfile != nil
```
Please add a comment explaining the implications of `mountsLockfile`.

saschagrunert replied:
Added comment as suggested :+1:

nalind commented on store.go:
```diff
@@ -2705,7 +2764,7 @@ func (s *store) Changes(from, to string) ([]archive.Change, error) {
 
 func (s *store) DiffSize(from, to string) (int64, error) {
-	lstore, err := s.LayerStore()
+	lstore, err := s.ROLayerStore()
```
Pretty sure this needs to be a writeable store, in order for `Changes()` to be able to do its work reliably.  Attempts to compute this information on read-only stores have a good chance of erroring out on read-only filesystems.

saschagrunert replied:
Alright, changed it to `RWLayerStore()`

nalind commented on store.go:
```diff
@@ -1745,7 +1804,7 @@ func (s *store) ImageSize(id string) (int64, error) {
 
 func (s *store) ContainerSize(id string) (int64, error) {
-	lstore, err := s.LayerStore()
+	lstore, err := s.ROLayerStore()
```
Pretty sure this needs to be a writeable store, because it uses `DiffSize()`.  We probably don't need to bother with any read-only stores in here, since we should only be finding container read-write layers in the read-write store, which is the only place we can create them.

saschagrunert replied:
I changed it to `RWLayerStore()` and removed the ROLayers. PTAL

nalind commented on store.go:
```diff
@@ -2920,6 +2979,24 @@ func (s *store) Layers() ([]Layer, error) {
+	if err != nil {
+		return nil, err
+	}
+	layers, err := lstore.Layers()
```
How does `store.Layers()` notice changes in the primary layers list?

saschagrunert replied:
I think they're independent right?

nalind commented on layers.go:
```diff
@@ -491,7 +491,10 @@ func (r *layerStore) saveMounts() error {
 }
 
-func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Driver) (LayerStore, error) {
+// newLayerStore creates a new layer storage baed on the provided directories
```
Typo: `baed`

saschagrunert replied:
Fixed :+1:

nalind commented:
A couple of methods seem to need to be using read-write layer stores, and I'm not sure consulting read-only stores for information about container read-write layers ever made sense.

saschagrunert commented:
> A couple of methods seem to need to be using read-write layer stores, and I'm not sure consulting read-only stores for information about container read-write layers ever made sense.

Alright I added another commit on top to see if it has any bad influence on the tests. I will squash them if they're fine.

rhatdan commented:
Could you vendor this into podman and/or crio and make sure it passes their CI systems.

saschagrunert commented:
> Could you vendor this into podman and/or crio and make sure it passes their CI systems.

Sure, here are the PRs:
- https://github.com/containers/libpod/pull/5706
- https://github.com/cri-o/cri-o/pull/3521

Let's see if I can get them green

saschagrunert commented:
We have a server error on the Ubuntu instances when it comes to download some packages.

saschagrunert commented:
https://github.com/containers/libpod/pull/5706 should get green now with those vendored changes.

saschagrunert commented:
The test failures seem not to be related to my change:
```
[+0316s] Hit:11 http://ppa.launchpad.net/projectatomic/ppa/ubuntu disco InRelease
[+0316s] Reading package lists...
[+0316s] E: The repository 'http://us-central1.gce.archive.ubuntu.com/ubuntu disco Release' no longer has a Release file.
[+0316s] E: The repository 'http://us-central1.gce.archive.ubuntu.com/ubuntu disco-updates Release' no longer has a Release file.
[+0316s] E: The repository 'http://us-central1.gce.archive.ubuntu.com/ubuntu disco-backports Release' no longer has a Release file.
[+0316s] E: The repository 'http://security.ubuntu.com/ubuntu disco-security Release' no longer has a Release file.
[+0316s] ##### (failed with exit: 100)
```

Is there anything we can change in the CI to choose other servers?

rhatdan commented:
@cevich WDYT? ^^

cevich commented:
> Is there anything we can change in the CI to choose other servers?

@rhatdan @saschagrunert Sigh, these are cooked-in our Ubuntu images in GCE before we even get them.  Nevertheless, it is possible to override these and/or otherwise control the repository config. in `./contrib/cirrus/packer/ubuntu_setup.sh` (run during VM image creation).  Unf. I'm not well versed in Ubuntu|Debian, so don't know how to actually affect these changes.  If someone wants to figure it out, I'm happy to help consult on the building-part.  [Anyone can do it, as documented.](https://github.com/containers/libpod/blob/master/contrib/cirrus/README.md#test_build_cache_images_task-task)

saschagrunert commented:
That's totally strange, are they currently in a release process for ubuntu disco? Because this does not work for me as well:

```
> podman run -it ubuntu:disco apt-get update
Ign:1 http://security.ubuntu.com/ubuntu disco-security InRelease
Err:2 http://security.ubuntu.com/ubuntu disco-security Release
  404  Not Found [IP: 91.189.88.152 80]
Ign:3 http://archive.ubuntu.com/ubuntu disco InRelease
Ign:4 http://archive.ubuntu.com/ubuntu disco-updates InRelease
Ign:5 http://archive.ubuntu.com/ubuntu disco-backports InRelease
Err:6 http://archive.ubuntu.com/ubuntu disco Release
  404  Not Found [IP: 91.189.88.152 80]
Err:7 http://archive.ubuntu.com/ubuntu disco-updates Release
  404  Not Found [IP: 91.189.88.152 80]
Err:8 http://archive.ubuntu.com/ubuntu disco-backports Release
  404  Not Found [IP: 91.189.88.152 80]
Reading package lists... Done
E: The repository 'http://security.ubuntu.com/ubuntu disco-security Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://archive.ubuntu.com/ubuntu disco Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://archive.ubuntu.com/ubuntu disco-updates Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://archive.ubuntu.com/ubuntu disco-backports Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```

saschagrunert commented:
@rhatdan it looks like Ubuntu Disco reached end of life:
https://lists.ubuntu.com/archives/ubuntu-announce/2020-January/000253.html

@cevich do we have any chance to switch the distribution here?

cevich commented:
> @cevich do we have any chance to switch the distribution here?> 

Oh, thanks for finding that, I'm working on getting this into containers/buildah in https://github.com/containers/buildah/pull/1971 I can try a similar change here...

cevich commented:
... https://github.com/containers/storage/pull/603

saschagrunert commented:
Rebased on top of the latest master branch

saschagrunert commented:
@rhatdan @nalind @vrothberg please take a look if you got some cycles.

saschagrunert commented:
@rhatdan @nalind @vrothberg do you think we can move forward with this enhancement?

akostadinov commented:
My +1 here. It's unreasonable to wait half an hour while images are being pushed before any operations can be performed. While if I have some container already running, it doesn't block pushing (thankfully).

rhatdan commented:
@nalind @giuseppe @vrothberg Interested in taking another look at this one?  Would be nice to get this functionality in, if possible.

vrothberg commented:
Apologies, I will do another review this week. Podman v2 and other high-priority work items were eating up my time.

vrothberg commented on store.go:
```diff
@@ -815,35 +822,88 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
 // used by the Store.  Accessing this store directly will bypass locking and
 // synchronization, so it is not a part of the exported Store interface.
 func (s *store) LayerStore() (LayerStore, error) {
```
Maybe we can mention that `LayerStore` is deprecated and `RWLayerStore` should be used? Or something similar to make clear that it's just an alias.

saschagrunert replied:
Sure, added a deprecation notice.

saschagrunert commented:
@rhatdan do you know where the CI results went down? I think only `lint`, `meta` and `vendor` does not seem enough reports. :/

cevich commented:
@saschagrunert that's a question for me...My hunch is it's related to something breaking here: https://github.com/containers/storage/blob/master/.cirrus.yml#L59-L74

This seems to also be the case for builds running on the master branch :confounded: Unfortunately there's no easy way (that I'm aware of) to test for missing jobs from within Cirrus-CI.  The only mechanism I'm aware of is to add a branch-protection rule...so it's up to @rhatdan if we want that (I can add them easy enough).

Curiously, in a different project, under a different config entirely, I was beating myself up a few weeks ago because `only_if` wasn't applying to my matrix.  Because of that, I'm wondering if there is some breakage on the Cirrus-CI side that is causing this.  I will send them an e-mail and ask..

cevich commented:
Okay, Cirrus-CI support helped track down the problem to a change on their end and a misconfiguration in our yaml (we were depending on broken behavior).  Working up a fix, and debugging with them over here: https://github.com/containers/storage/pull/677

saschagrunert commented:
> Okay, Cirrus-CI support helped track down the problem to a change on their end and a misconfiguration in our yaml (we were depending on broken behavior). Working up a fix, and debugging with them over here: #677

Thank you for helping on a fix @cevich!

saschagrunert commented:
> Okay, Cirrus-CI support helped track down the problem to a change on their end and a misconfiguration in our yaml (we were depending on broken behavior). Working up a fix, and debugging with them over here: #677

Thank you for helping on a fix @cevich!

cevich commented:
Update: Turns out actually fixing the underlying problem on the CIrrus-CI side was more trouble than expected.  They have reverted the change that caused the breakage, and so my PR has become mostly just additional documentation.  I've also added a `success` accumulator task, and will mark it as *required* in github.  This will help in similar situations to prevent merges when fewer than all expected test results are present.

No rebase is necessary here, simply re-running the tests should function normally now.  Sorry for any confusion this caused.

cevich commented:
Correction: Re-running tasks won't work, we need an entirely new build to be created.  Make some small inconsequential change somewhere and `--force` push.  I've confirmed in my "fix" PR this will cause the tests to run properly.

saschagrunert commented:
> Correction: Re-running tasks won't work, we need an entirely new build to be created. Make some small inconsequential change somewhere and `--force` push. I've confirmed in my "fix" PR this will cause the tests to run properly.

Looks like that the tests are running again, thanks!

saschagrunert commented:
At least this PR seems green now.

rhatdan commented:
@saschagrunert This PR has merge issues.  Needs a rebase.

saschagrunert commented:
> @saschagrunert This PR has merge issues. Needs a rebase.

I rebased on top of the version Podman vendors to have a small diff. I'll rebase on top of the master once the issues with the Podman CI is solved :+1:

TomSweeneyRedHat commented:
Oh my, this has all green test buttons atm....

TomSweeneyRedHat commented:
The changes LGTM, but I definitely want a thorough @nalind review on this one.

saschagrunert commented:
CI seems partly failing again because of:
```
[+0212s] Errors were encountered while processing:
[+0212s]  grub-efi-amd64-signed
[+0214s] E: Sub-process /usr/bin/dpkg returned an error code (1)
```

rhatdan commented:
@cevich PTAL

rhatdan commented:
@kolyshkin I would love to have  you review this PR.

TomSweeneyRedHat commented on store.go:
```diff
@@ -826,36 +838,90 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
+func (s *store) ROLayerStore() (ROLayerStore, error) {
+	s.resetIfNeeded()
+
+	// Memoized read-only layer store
```
nit of a nit
```suggestion
	// Memorized read-only layer store
```

rhatdan replied:
Memoized is a word.
https://en.wikipedia.org/wiki/Memoization

saschagrunert replied:
Yep, that was intended :+1:

saschagrunert commented:
I'm kind of in a trial and error mode now with the Podman PR https://github.com/containers/podman/pull/7603. I need the integration/system tests there to see if it works, because the issue is not reproducible on my machine. 

I'll keep you posted, this can take some iterations.

saschagrunert commented:
The Podman PR as well as this one is now green, but I have to verify that my changes logically make sense…
https://github.com/containers/podman/pull/7603

rhatdan commented:
@saschagrunert Just to move the boudary a little further, could you try a PR against CRI-O to see how that goes.

saschagrunert commented:
> @saschagrunert Just to move the boudary a little further, could you try a PR against CRI-O to see how that goes.

Yep, opened the PR there: https://github.com/cri-o/cri-o/pull/4339

rhatdan commented:
@vrothberg @nalind @TomSweeneyRedHat @mtrmac @mrunalp This PR is now passing tests in Podman and CRI-O, I believe it is time to give it another review.

akostadinov commented:
@saschagrunert , is your work intended to fix issues like:
https://github.com/containers/podman/issues/8667

kolyshkin commented on layers.go:
```diff
@@ -491,7 +491,10 @@ func (r *layerStore) saveMounts() error {
-func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Driver) (LayerStore, error) {
+// newLayerStore creates a new layer storage based on the provided directories
+// and selected driver. The `withMountsLock` bool indicates if we want to
+// enable mount locking, which is necessary to have write access to the storage
```
nit: missing period at EOL

saschagrunert replied:
Fixed

kolyshkin commented on store.go:
```diff
@@ -826,36 +838,90 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
+}
+
+// getLayerPaths retrieves the run directory (`rlpath`) the layer directory
+// (`glpath`) and the currently configured storage driver for the store
```
We probably should not use github markup (backticks) in godoc comments.

Also, no period at EOL.

saschagrunert replied:
Fixed

kolyshkin commented on store.go:
```diff
@@ -826,36 +838,90 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
+	if err != nil {
+		return "", "", nil, err
+	}
+	driverPrefix := s.graphDriverName + "-"
```
I'd say have
```suggestion
layers := s.graphDriverName + "-" + "layers"
```
and use it below.

saschagrunert replied:
Fixed as suggested

kolyshkin commented on store.go:
```diff
@@ -826,36 +838,90 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
+
+// getLayerPaths retrieves the run directory (`rlpath`) the layer directory
+// (`glpath`) and the currently configured storage driver for the store
+func (s *store) getLayerPaths() (rlpath, glpath string, driver drivers.Driver, err error) {
```
This looks more like `makeLayerPaths`...

saschagrunert replied:
Renamed to `makeLayerPaths`

kolyshkin commented on store.go:
```diff
@@ -1758,22 +1823,15 @@ func (s *store) ImageSize(id string) (int64, error) {
-			}
+	lstore.RLock()
+	defer lstore.Unlock()
+	if modified, err := lstore.Modified(); modified || err != nil {
```
I'd reverse the order of checks here, ie
```suggestion
	if modified, err := lstore.Modified(); err != nil || modified {
```

as you're not supposed to check `modified` in case `err != nil`.

saschagrunert replied:
Fixed :+1:

kolyshkin commented on store.go:
```diff
@@ -3049,11 +3121,27 @@ func (s *store) Containers() ([]Container, error) {
 
-func (s *store) Layer(id string) (*Layer, error) {
-	lstore, err := s.LayerStore()
+func (s *store) Layer(id string) (layer *Layer, err error) {
```
Looks like there's no need for named return parameters here.

saschagrunert replied:
Fixed :+1:

saschagrunert reviewed (commented):
@kolyshkin I fixed your mentioned review points. PTAL

akostadinov commented:
@saschagrunert , sorry for double posting but I think it is important to know whether your PR is suppose to fix issues like containers/podman#8667

If not, then could you clarify what use cases should be considered supported?

mtrmac commented on layers.go:
```diff
@@ -524,27 +530,6 @@ func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Dri
 }
 
-func newROLayerStore(rundir string, layerdir string, driver drivers.Driver) (ROLayerStore, error) {
-	lockfile, err := GetROLockfile(filepath.Join(layerdir, "layers.lock"))
```
`newLayerStore` as a replacement for this will always create a RW lockfile. Is that desirable? It changes semantics of `r.lockfile.RecursiveLock` == `r.RecursiveLock`, which seems worrisome.

OTOH https://github.com/containers/storage/blob/9f889830f10075aceb6b925a97927e77b837d0a7/pkg/lockfile/lockfile.go#L92-L100 does not support having both kinds of lock for a single path.

mtrmac commented on layers.go:
```diff
@@ -1470,7 +1455,8 @@ func (r *layerStore) Modified() (bool, error) {
 func (r *layerStore) IsReadWrite() bool {
-	return r.lockfile.IsReadWrite()
+	// the mountsLockfile is only set for read-write enabled stores.
+	return r.lockfile.IsReadWrite() && r.mountsLockfile != nil
```
`r.lockfile.IsReadWrite` is always true now, AFAICS.

mtrmac commented on store.go:
```diff
@@ -827,36 +839,90 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
+// Deprecated: Use RWLayerStore() instead
 func (s *store) LayerStore() (LayerStore, error) {
-	s.graphLock.Lock()
-	defer s.graphLock.Unlock()
```
Previously, this lock protected all of `LayerStore`, including `s.getGraphDriver`; now the lock is scoped to `resetIfNeeded`; why is that sufficient? (It might be, but it’s not at all obvious to me — the lock’s scope is not documented.)

mtrmac commented on layers.go:
```diff
@@ -491,7 +491,10 @@ func (r *layerStore) saveMounts() error {
 
-func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Driver) (LayerStore, error) {
+// newLayerStore creates a new layer storage based on the provided directories
+// and selected driver. The `withMountsLock` bool indicates if we want to
```
Is it really `withMountsLock` or `readWrite`? The latter seems semantically more useful, and might e.g. allow skipping the `MkdirAll` calls below.

mtrmac commented on store.go:
```diff
@@ -582,6 +593,7 @@ type store struct {
 	autoNsMaxSize   uint32
 	graphDriver     drivers.Driver
 	layerStore      LayerStore
```
Name this `rwLayerStore`? Possibly document the fields, even if they are “a private cache for $method”.

mtrmac commented on store.go:
```diff
@@ -582,6 +593,7 @@ type store struct {
 	graphDriver     drivers.Driver
 	layerStore      LayerStore
+	roLayerStore    ROLayerStore
 	roLayerStores   []ROLayerStore
```
Suppose the user calls:
- `s.Names(foo)` // Populates `roLayerStore`, reads from disk
- `s.PutLayer(…)` // Populates `layerStore`, reads + writes to disk
- `s.Names(foo)`

What happens now? AFAICS:
- `s.graphLock.TouchedSince` is false (no `.Shutdown*`), so the cached `roLayerStore` is used
- The RO and RW layer stores share lock objects ( https://github.com/containers/storage/blob/9f889830f10075aceb6b925a97927e77b837d0a7/pkg/lockfile/lockfile.go#L92-L100 ), so regardless of any `.Touch()` calls on locks, `.Modified()` will be `false`
- The RO layer store will continue reading from its cached data, regardless of modifications by the RW store.

(This is only by reading code I’m unfamiliar with, so I may well be missing something or making a stupid mistake.)

mtrmac reviewed (commented):
Overall, as a stranger to the codebase, I’d appreciate fairly detailed comments about the design: What do the various locks protect, what is authoritative data vs. caches, and so on and so on.

- The concurrent maintenance of two stores seems risky to me, see elsewhere. Wouldn’t it make sense to only have an one-directional “upgrade”, where the `store` either has a read-only layer store, or replaces it with a read-write, but never creates a read-write one again?
- Having two stores means two caches = two reads of the data. Is that really worth it?
- If there has to be a separate `ShutdownRO` and the caller is expected to call it to benefit from it (and the code breaks if the caller calls RO after updates!), this is not completely transparent _anyway_. How are the library callers going to use it, then? If the code that creates/shuts down the `storage.Store` must know whether the store is RO or RW, it would seem _much_ cleaner to me to have that code _create_ the `storage.Store` in a RO or RW mode (and then c/storage itself can be in control of its invariants, e.g. refuse to do any write operations on a RO store). (Yes, I have noticed that this was considered about a year ago; as I said at the start, I’m a stranger to the code base and my opinion should not be authoritative — I can’t really understand the CRI-O argument in there, for example.)

saschagrunert commented:
> @saschagrunert , sorry for double posting but I think it is important to know whether your PR is suppose to fix issues like [containers/podman#8667](https://github.com/containers/podman/issues/8667)
> 
> If not, then could you clarify what use cases should be considered supported?

The main use case was to be able to list container images while another push operation in in progress. I think containers/podman#8667 could be solved when locking by layer and not the whole storage. But as folks already mentioned, this would mean a architectural rewrite of containers/storage and I'm not sure if we will get to this point anytime soon.

saschagrunert commented:
> * The concurrent maintenance of two stores seems risky to me, see elsewhere. Wouldn’t it make sense to only have an one-directional “upgrade”, where the `store` either has a read-only layer store, or replaces it with a read-write, but never creates a read-write one again?

I think if we put API compatibility aside for now, then we could create a store which is ready-only locked per default. Just some operations should then be able to lock the certain parts of the store. For example, storing a layer should only RW lock that part of the store. Other operations (read-only or read-write which do not affect the WIP layers) should be still possible.

The API compatibility is the main burden here. c/storage is designed to cleanup the layers after storage shutdown, which will always lock the complete layer store for example.

> * Having two stores means two caches = two reads of the data. Is that really worth it?

No I don't think so and agree with your concern. The only reason to implement it like this is to stay compatible to current consumers of this project.

> * If there has to be a separate `ShutdownRO` and the caller is expected to call it to benefit from it (and the code breaks if the caller calls RO after updates!), this is not completely transparent _anyway_. How are the library callers going to use it, then?

Manually, library consumers have to decide on their own if they want to shutdown with cleanup or without. For example a `podman ps` would not have a need for a layer cleanup. I know that is a major drawback of the proposal.

---

All in all I think we will have a hard time keeping the current API stable while improving the locking. We could probably start with a parallel implementation (which breaks the API) and integrate the enhancements mentioned by @mtrmac. 

If we speak only about refactoring the layer store, then the main strategy would be:
1. Redesign a locker which is able to lock a single layer
1. Use only one layer store for c/storage which is using the new locking (read-only per default)
1. Decide which operations will automatically and partially rw lock the layer store
1. Cleanup the store based on its usage (aka track with a flag if the store needs cleanup).

What do you think?

vrothberg commented:
> Redesign a locker which is able to lock a single layer

Fine-grained locking is _very_ difficult to get right and a common source of concurrency bugs.  Having a per-layer lock, for instance, is subject to ABBA dead locks (assuming a process is allowed to lock multiple layers at once).

One thing we must consider as well is backward compatibility. If we fundamentally change how c/storage works, older and newer versions cannot run on the same machine.

I prefer to keep the coarse-grained locks unless there is no alternative.

There may still be more space for optimization (see https://github.com/containers/storage/issues/420#issuecomment-523894539) which does not entirely rule out lock monopolization.

mtrmac commented:
> I think if we put API compatibility aside for now, then we could create a store which is ready-only locked per default. Just some operations should then be able to lock the certain parts of the store. For example, storing a layer should only RW lock that part of the store. Other operations (read-only or read-write which do not affect the WIP layers) should be still possible.

What happens with the current consistency fix-ups that write data on load (see `shouldSave` in `layerStore.Load` and `imageStore.Load` at least)? Assuming we don’t somehow modify the code to transparently read inconsistent data everywhere (which might be possible, but seems difficult to make reliable), any operation _can_ obtain a write lock even if the caller “just wants to read something”. That suggests to me that all of this probably _could_ be made API-transparent: keep the existing API, and just lock only for reading if necessary… 

… except that, well, that’s what the code is _already_ doing: `store.go` calls `store.RLock()` for read-only operations.

So… what’s the _actual failure path_ this PR is trying to address? That `layerStore.Load()` uses `r.lockfile.Lock()`, which is read-write, and that always blocks on a concurrent push? Couldn’t that be a fairly local fix, where `layerStore.Load()` tries to load with a RO lock only, and on consistency failure it gives that lock up, and tries again with RW lock?

vrothberg commented:
I opened https://github.com/containers/image/pull/1167 which is tackling the issue with a different perspective.

Ultimately, solving the issue in storage is favorable but as shown here, it's very difficult. If callers need to worry about executions paths reading or writing data is likely causing future regressions as it's hard to get right. 

https://github.com/containers/image/pull/1167 makes pushing a bit slower but it solves the problem of being network bound until unlocking the storage.
